File: src/main/java/core/db/StatementCache.java
package core.db;

import core.util.HOLogger;

import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.time.Instant;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class StatementCache {
    private final ConnectionManager connectionManager;

    private boolean cachedEnabled = true;

    private final Map<String, PreparedStatement> cache = Collections.synchronizedMap(new HashMap<>());

    private final Map<String, CachedStatementStats> statementStats = Collections.synchronizedMap(new HashMap<>());

    public StatementCache(ConnectionManager connectionManager) {
        this.connectionManager = connectionManager;
    }

    public void setCachedEnabled(boolean enabled) {
        cachedEnabled = enabled;
    }

    private PreparedStatement getFromCache(String query) {
        if (cachedEnabled) {
            PreparedStatement statement = cache.get(query);
            if (statement != null) {
                CachedStatementStats stats = statementStats.get(query);
                statementStats.put(query, new CachedStatementStats(stats.created, Instant.now(), stats.count + 1));
                return statement;
            }
        }
        return null;
    }

    private PreparedStatement createStatement(String query) {
        PreparedStatement statement = null;
        try {
            statement = connectionManager.connection.prepareStatement(query);
            if (cachedEnabled) {
                cache.put(query, statement);
                statementStats.put(query, new CachedStatementStats(Instant.now(), Instant.now(), 1));
            }
        } catch (SQLException e) {
            HOLogger.instance().error(StatementCache.class, "Error creating statement: " + query
                    + "\n Error: " + e.getMessage());
        }

        return statement;
    }

    public PreparedStatement getPreparedStatement(String query) {
        PreparedStatement statement = getFromCache(query);
        if (statement == null) {
            statement = createStatement(query);
        }
        return statement;
    }


    record CachedStatementStats(Instant created, Instant lastAccessed, int count) {}
}

File: src/main/java/core/db/StatementCache.kt
package core.db

import core.util.HOLogger
import java.sql.PreparedStatement
import java.sql.SQLException
import java.time.Instant
import java.util.*

/**
 * Cache for [PreparedStatement]s instances.
 *
 *
 * This cache tracks statistics about the various prepared statements:
 *
 *  * Creation timestamp,
 *  * Last access timestamp,
 *  * NUmber of accesses.
 *
 *
 *
 * The cache can be disabled by setting `cachedEnabled` to `false`.  When the cache
 * is disabled, the existing entries are closed and evicted, the stats dumped and cleared.  The cache can
 * be enabled or disabled via JMX in development mode.  By default, the cache is on.
 */
class StatementCache(private val connectionManager: ConnectionManager) {
    var cachedEnabled = true
        set(enabled) {
            field = enabled
            HOLogger.instance().info(StatementCache::class.java, "Cache enabled = $enabled")
            if (!field) {
                clearCache()
            }
        }

    private val cache = Collections.synchronizedMap(HashMap<String, PreparedStatement?>())

    val statementStats: MutableMap<String, CachedStatementStats> = Collections.synchronizedMap(HashMap())

    private fun getFromCache(query: String): PreparedStatement? {
        if (cachedEnabled) {
            val statement = cache[query]
            if (statement != null) {
                val stats = statementStats[query]
                statementStats[query] = CachedStatementStats(stats!!.created, Instant.now(), stats.count + 1)
                return statement
            }
        }
        return null
    }

    private fun createStatement(query: String): PreparedStatement? {
        var statement: PreparedStatement? = null
        try {
            statement = connectionManager.connection.prepareStatement(query)
            if (cachedEnabled) {
                cache[query] = statement
                statementStats[query] = CachedStatementStats(Instant.now(), Instant.now(), 1)
            }
        } catch (e: SQLException) {
            HOLogger.instance().error(
                StatementCache::class.java, """Error creating statement: $query
 Error: ${e.message}"""
            )
        }
        return statement
    }

    fun getPreparedStatement(query: String): PreparedStatement? {
        var statement = getFromCache(query)
        if (statement == null) {
            statement = createStatement(query)
        }
        return statement
    }

    fun clearCache() {
        for ((key, value) in cache) {
            try {
                value!!.close()
            } catch (e: SQLException) {
                HOLogger.instance().error(
                    StatementCache::class.java,
                    """Error closing prepared statement: $key
 ${e.message}"""
                )
            }
        }
        cache.clear()
        dumpStats()
        statementStats.clear()
    }

    fun dumpStats() {
        for ((key, value) in statementStats) {
            HOLogger.instance().info(StatementCache::class.java, "$key: $value")
        }
    }

    @JvmRecord
    data class CachedStatementStats(val created: Instant, val lastAccessed: Instant, val count: Int)
}
